File: tests/benchmarks/resolution/fixtures/haskell/Repository.hs
module Repository
  ( save
  , findById
  , delete
  , count
  ) where

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map

type UserId = String
type UserRecord = (String, String, Int)

save :: UserId -> UserRecord -> Map UserId UserRecord -> Map UserId UserRecord
save uid record store = Map.insert uid record store

findById :: UserId -> Map UserId UserRecord -> Maybe UserRecord
findById uid store = Map.lookup uid store

delete :: UserId -> Map UserId UserRecord -> Map UserId UserRecord
delete uid store = Map.delete uid store

count :: Map UserId UserRecord -> Int
count store = Map.size store

File: tests/benchmarks/resolution/fixtures/erlang/repository.erl
-module(repository).
-export([init/0, save/2, find_by_id/2, delete/2, list_all/1]).

init() ->
    ets:new(user_store, [set, named_table, public]).

save(User, Table) ->
    Id = maps:get(id, User),
    ets:insert(Table, {Id, User}),
    User.

find_by_id(Id, Table) ->
    case ets:lookup(Table, Id) of
        [{_, User}] -> {ok, User};
        [] -> {error, not_found}
    end.

delete(Id, Table) ->
    ets:delete(Table, Id),
    ok.

list_all(Table) ->
    ets:tab2list(Table).
